src/bootstrap/metrics.rs

```diff
@@ -131,7 +152,20 @@ impl BuildMetrics {
         // Some of our CI builds consist of multiple independent CI invocations. Ensure all the
         // previous invocations are still present in the resulting file.
         let mut invocations = match std::fs::read(&dest) {
-            Ok(contents) => t!(serde_json::from_slice::<JsonRoot>(&contents)).invocations,
+            Ok(contents) => {
+                // We first parse just the metadata_version field to have the check succeed even if
+                // the rest of the contents are not valid anymore.
+                let version: OnlyMetadataVersion = t!(serde_json::from_slice(&contents));
+                if version.metadata_version == CURRENT_METADATA_VERSION {
+                    t!(serde_json::from_slice::<JsonRoot>(&contents)).invocations
+                } else {
+                    println!(
+                        "warning: overriding existing build/metrics.json, as it's not \
+                         compatible with build metrics format version {CURRENT_METADATA_VERSION}."
+                    );
```

Comment:
I am confused - why would we need to support both versions at once? Is `metrics.json` somehow cached between different PRs?

Reply:
So, there are two sides to this:

* In general, I added the metadata version number as the metrics are expected to be consumed by tools managed out of tree (like Mark's scripts to generate graphs, infra's script to measure CI usage, or my Ferrocene tooling), and it's useful to signal those tools when a breaking change happens.

* As you pointed out CI does not cache the build metrics between different PRs, but if someone sets `build.metrics = true` locally the `build/metrics.json` file will continue to be updated, and it'd result in a confusing serde error message once a breaking change happens. At least I have metrics enabled in some worktrees, mostly when I need to make changes to the metrics and I forget to turn it off. Gracefully handling this is not much code, and provides better behavior for interactive users.

Reply:
:+1: that makes sense - can you add a comment in the code mentioning those two use cases?

Reply:
Done!